lib/data_magic/document_builder.rb

```diff
@@ -0,0 +1,140 @@
+require './lib/expression'
+
+module DataMagic
+  module DocumentBuilder
+    class << self
+      def logger
+        DataMagic::Config.logger
+      end
+
+      # parse a row from a csv file, returns a nested document
+      # row: a hash  { field => value } where all values are strings
+      # fields: column_name => field_name
+      # config: DataMagic.Config instance for dictionary, column types, NULL
+      def parse_row(row, fields, config, options={}, additional=nil)
```

Comment:
This method is quite hard to read. It keeps reusing and mutating `row`, which then changes to `doc`, which then gets mutated more. At some point in our Copious Free Time we should refactor this code to be a Document/Row object, and the mutation can be more direct with more meaningful method names. As it is now, it's quite hard to work out what's going on and why.